doc/man/man1/pabopto2bsdf.1
.\" RCSid "$Id: pabopto2bsdf.1,v 1.11 2025/05/25 17:57:18 greg Exp $"
.TH PABOPTO2BSDF 1 2/24/2021 RADIANCE
.SH NAME
pabopto2bsdf - convert BSDF measurements to a scattering interpolant representation
.SH SYNOPSIS
.B pabopto2bsdf
[
.B \-t
][
.B "\-n nproc"
][
.B "\-s symmetry"
][
.B "\-g angle | 'A'"
]
.B "meas1 meas2 .."
.SH DESCRIPTION
.I Pabopto2bsdf
takes two or more pab-opto
.I Mountain
files, each nominally containing
different incident beam angles or sampling patterns,
and produces a Scattering Interpolant Representation (SIR)
on the standard output for further processing.
The binary SIR contains a Radial Basis Function fitting
each incident BSDF data file
and a "transport plan" matrix for each pair of neighboring RBF
directions in a spherical Delaunay mesh.
.PP
The SIR provides a complete 4-dimensional
BSDF description that may be resampled for other
formats such as Klems and tensor tree.
However, a separate run of
.I pabopto2bsdf
is needed to produce an SIR for each
incident and scattered hemisphere pair.
At most, there will be 4 such hemisphere pairs for
front reflection, back reflection, front transmission,
and back transmission.
Theoretically, only one transmission direction is required,
but it is often safest to measure both if they are to
be used in a simulation.
(See
.I bsdf2klems(1)
and
.I bsdf2ttree(1)
for details.
The
.I bsdf2rad(1)
and
.I bsdfview(1)
tools are also useful for visualizaing SIR and XML files.)
.PP
The
.I pabopto2bsdf
.I \-t
option reverses the assumed sample orientation front-to-back,
and is discussed below under the "#intheta" header entry.
.PP
Multi-processing may be used to accelerate the program
on systems that support it via the
.I \-n
option.
.PP
BSDF symmetry may be specified with the
.I \-s
option, which is one of "isotropic", "quadrilateral",
"bilateral", "up", or "anisotropic".
Any of these may be abbreviated with as little as a single
letter, and case is ignored.
.PP
Normally,
.I pabopto2bsdf
will assume a BSDF symmetry from the incident phi angles provided.
If every input data file uses the same incident phi angle, the
BSDF is assumed to be "isotropic", or rotationally symmetric.
If input phi angles only cover one quarter of the incident hemisphere,
then the sample is assumed to have quadrilateral symmetry.
Similarly, half-hemisphere coverage implies "bilateral" symmetry,
although it is also compatible with "up" symmetry, which must be specified
on the command line.
The difference is crucial.
Similar to quadrilateral symmetry, bilateral symmetry is "mirrored,"
meaning that the sample material looks identical when viewed in a mirror.
However, "up" symmetry means that the sample looks the same when
rotated by 180-degree (upside-down), but does not look the same in a mirror.
The "up" symmetry was a late addition, and involves rotating and copying the
input data, treating the result as anisotropic.
It is therefore less efficient, and should only be used when necessary.
Finally, if the incident hemisphere is fully covered, the final BSDF
is anisotropic.
.PP
If a
.I \-s
symmetry option is specified and it does not agree with the input
data provided, an error message is issued and no output is produced.
Note that only the "up" and "bilateral" symmetry options have
identical input coverage, so this is the only time the
.I \-s
option must be specified if the default mirroring is not appropriate.
.PP
If a
.I \-g
option is present, it will cull scattered measurements that are nearer
to grazing than the given angle in degrees.
If the word "auto" (which can be abbreviated as 'a' or 'A') is given
instead of an angle, then the near-grazing angle will be determined
by the lowest incident angle measurement present in the input data.
This is sometimes necessary to eliminate noise and edge effects that
some measurements exhibit near grazing.
.PP
The
.I Mountain
program, written by Peter Apian-Bennewitz, stores data taken by 
his pg2 goniophotometer in separate
BSDF scattering files for each incident angle, beginning with a header
whose lines each start with a pound sign ('#').
Some header settings require colons and others do not, as indicated below.
The
.i pabopto2bsdf
program understands the following lines from each header and ignores
the rest:
.TP
.BR #sample_name
A double-quoted string containing the name associated with this sample.
If input files contain different sample names, the final sample name read
will be the one passed to the SIR output.
.TP
.BR #format:
The data format, typically one of "theta phi DSF" or "theta phi BSDF".
These differ only in their inclusion of a cosine factor.
The word "BRDF" or "BTDF" is accepted in place of "BSDF".
Any other specification or a format missing generates an error.
.TP
.BR #intheta
The incident theta (polar) angle in degrees, measured from the sample's
surface normal.
Theta values should be between 0 and 180, where values less than 90
are considered incident to the "front" side of the sample, and
theta values greater than 90 are incident to the "back" side in
the standard coordinate system.
Notions of "front" and "back" may be reversed using the
.I -t
option if desired.
.TP
.BR #inphi
The incident phi (azimuthal) angle in degrees counter-clockwise as
seen from the "front" side of the sample.
.TP
.BR #incident_angle
The incident theta and phi angles are each given in this header
line, offered as an alternative to separate "#intheta" and "#inphi"
angles.
The interpretation is the same as above.
.TP
.BR #upphi
If present, this phi angle that corresponds to
the sample "up" orientation.
By default, it is assumed to be 0, meaning that "up"
is phi=0.
To get the standard RADIANCE coordinates for BSDFs, "#upphi" should
be set to 90 (degrees).
.TP
.BR #colorimetry:
Two colorimetry values are currently understood: "CIE-Y" and "CIE-XYZ".
The default "CIE-Y" colorimetry
takes each DSF or BSDF value as photometric.
If "CIE-XYZ" is specified, then the DSF or BSDF values must be triplets
corresponding to CIE XYZ values.
Such files are typically produced by the
.I pabopto2xyz(1)
tool rather than
.I Mountain,
directly.
.PP
The BSDF scattering data follows the header in unspecified order,
where each line in the file
contains the scattered theta and phi angles measured in the same
coordinate system as incident theta and phi, followed by the DSF
or BSDF value, which may either be a single photometric quantity
for "CIE-Y" colorimetry or a triplet if the colorimetry is "CIE-XYZ".
A minimal incident BSDF data file might contain:
.sp
.nf
#incident_angle 82.5 180
#format: theta phi DSF
84.968 125.790 0.009744
84.889 125.610 0.007737
84.805 125.427 0.008569
 ...
.fi
.sp
The above header is equivalent to the more complete version below:
.sp
.nf
#format: theta phi DSF
#incident_angle 82.5 180
#intheta 82.5
#inphi 180
#upphi 0
#colorimetry: CIE-Y
84.968 125.790 0.009744
84.889 125.610 0.007737
84.805 125.427 0.008569
 ...
.fi
.sp
The ordering of the header and data lines is unimportant,
but all header lines must precede all data lines in each input file.
.SH EXAMPLE
To generate an SIR file from a collection of transmission measurements
of a material with 180-degree symmetry using 4 processes:
.IP "" .2i
pabopto2bsdf -n 4 -s up f*_Tvis.txt > front_trans.sir
.PP
To combine this with front reflection measurements into a Klems BSDF file:
.IP "" .2i
pabopto2bsdf -n 4 -s up f*_Rvis.txt > front_refl.sir
.br
bsdf2klems front_trans.sir front_refl.sir > Klems_bsdf.xml
.SH NOTES
If the BSDF is being mirrored and there is no measured theta=0 incident
angle data file, this part of the distribution is filled in
by a special procedure.
This is important because there is no way to extrapolate missing
data at normal incidence.
.PP
The BSDF is extrapolated past the last measured theta angles towards
grazing using a constant value plus a single Gaussian lobe if one can
be reasonably fit to the near-grazing data.
This lobe will always be in the mirror direction in the case of
reflection, or the "through" direction in the case
of transmission.
The magnitude and width of this lobe is stored in the output header,
along with the constant value.
Both the lobe and the constant are neutral values, even with CIE-XYZ
colorimetry.
.PP
While there is no explicit handling of infrared or solar radiometry,
any single-channel BSDF will be created the same, and the final XML
file generated by
.I bsdf2klems
or
.I bsdf2ttree
can be edited to specify a different radiometry.
The interpolation process in
.I pabopto2bsdf
is not affected by this.
.PP
The standard BSDF coordinates in RADIANCE have the theta=0 direction
corresponding to the front-side surface normal.
The phi=0 direction points to the right as seen from the front, and
phi=90 degrees corresponds to the "up" orientation for the sample.
The same theta and phi are used for incoming and scattered angles,
so theta=180 is the opposite side surface normal.
This differs from the WINDOW, which use separate
coordinate systems for the front and the back.
To confuse things further, notions of "front" and "back" are
opposite in WINDOW and RADIANCE.
In RADIANCE, the normal of a window surface usually faces the
interior of a space, where the WINDOW convention uses "FRONT"
for the exterior of a building.
.PP
In the
.I genBSDF(1)
utility, the world coordinate system follows trigonometric
conventions with theta=0 aligning to the Z-axis,
the X-axis matches (theta,phi)=(90,0), and the Y-axis
corresponds to (theta,phi)=(90,90).
The latter is thought of as the "up" direction for the sample.
This usually needs to be rotated into position, since most
RADIANCE models use the Z-axis as the world "up" direction.
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
bsdf2klems(1), bsdf2rad(1), bsdf2ttree(1), bsdfview(1), genBSDF(1),
pabopto2xyz(1)


src/cv/pabopto2bsdf.c
#ifndef lint
static const char RCSid[] = "$Id: pabopto2bsdf.c,v 2.42 2022/11/21 18:28:24 greg Exp $";
#endif
/*
 * Load measured BSDF data in PAB-Opto format.
 * Assumes that surface-normal (Z-axis) faces into room unless -t option given.
 *
 *	G. Ward
 */

#define _USE_MATH_DEFINES
#include <stdlib.h>
#include <ctype.h>
#include <math.h>
#include "rtio.h"
#include "platform.h"
#include "paths.h"
#include "bsdfrep.h"
				/* global argv[0] */
char			*progname;

typedef struct {
	const char	*fname;		/* input file path */
	double		theta, phi;	/* input angles (degrees) */
	double		up_phi;		/* azimuth for "up" direction */
	int		igp[2];		/* input grid position */
	int		isDSF;		/* data is DSF (rather than BSDF)? */
	int		nspec;		/* number of spectral samples */
	long		dstart;		/* data start offset in file */
} PGINPUT;

PGINPUT		*inpfile;	/* input files sorted by incidence */

int		rev_orient = 0;	/* shall we reverse surface orientation? */

double		lim_graze = 0;	/* limit scattering near grazing (deg above) */

/* Compare incident angles */
static int
cmp_indir(const void *p1, const void *p2)
{
	const PGINPUT	*inp1 = (const PGINPUT *)p1;
	const PGINPUT	*inp2 = (const PGINPUT *)p2;
	int		ydif = inp1->igp[1] - inp2->igp[1];
	
	if (ydif)
		return(ydif);

	return(inp1->igp[0] - inp2->igp[0]);
}

/* Assign grid position from theta and phi */
static void
set_grid_pos(PGINPUT *pip)
{
	FVECT	dv;

	if (pip->theta <= FTINY) {
		pip->igp[0] = pip->igp[1] = grid_res/2 - 1;
		return;
	}
	dv[2] = sin(M_PI/180.*pip->theta);
	dv[0] = cos(M_PI/180.*pip->phi)*dv[2];
	dv[1] = sin(M_PI/180.*pip->phi)*dv[2];
	dv[2] = sqrt(1. - dv[2]*dv[2]);
	pos_from_vec(pip->igp, dv);
}

/* Prepare a PAB-Opto input file by reading its header */
static int
init_pabopto_inp(const int i, const char *fname)
{
	FILE	*fp = fopen(fname, "r");
	char	buf[2048];
	int	c;
	
	if (fp == NULL) {
		fputs(fname, stderr);
		fputs(": cannot open\n", stderr);
		return(0);
	}
	inpfile[i].fname = fname;
	inpfile[i].isDSF = -1;
	inpfile[i].nspec = 0;
	inpfile[i].up_phi = 0;
	inpfile[i].theta = inpfile[i].phi = -10001.;
				/* read header information */
	while ((c = getc(fp)) == '#' || c == EOF) {
		char	typ[64];
		if (fgets(buf, sizeof(buf), fp) == NULL) {
			fputs(fname, stderr);
			fputs(": unexpected EOF\n", stderr);
			fclose(fp);
			return(0);
		}
		if (sscanf(buf, "sample_name \"%[^\"]\"", bsdf_name) == 1)
			continue;
		if (sscanf(buf, "colorimetry: %s", typ) == 1) {
			if (!strcasecmp(typ, "CIE-XYZ"))
				inpfile[i].nspec = 3;
			else if (!strcasecmp(typ, "CIE-Y"))
				inpfile[i].nspec = 1;
			continue;
		}
		if (sscanf(buf, "format: theta phi %s", typ) == 1) {
			if (!strcasecmp(typ, "DSF"))
				inpfile[i].isDSF = 1;
			else if (!strcasecmp(typ, "BSDF") ||
					!strcasecmp(typ, "BRDF") ||
					!strcasecmp(typ, "BTDF"))
				inpfile[i].isDSF = 0;
			continue;
		}
		if (sscanf(buf, "upphi %lf", &inpfile[i].up_phi) == 1)
			continue;
		if (sscanf(buf, "intheta %lf", &inpfile[i].theta) == 1)
			continue;
		if (sscanf(buf, "inphi %lf", &inpfile[i].phi) == 1)
			continue;
		if (sscanf(buf, "incident_angle %lf %lf",
				&inpfile[i].theta, &inpfile[i].phi) == 2)
			continue;
	}
	inpfile[i].dstart = ftell(fp) - 1;
	fclose(fp);
	if (inpfile[i].isDSF < 0) {
		fputs(fname, stderr);
		fputs(": unknown format\n", stderr);
		return(0);
	}
	if ((inpfile[i].theta < -10000.) | (inpfile[i].phi < -10000.)) {
		fputs(fname, stderr);
		fputs(": unknown incident angle\n", stderr);
		return(0);
	}
	if (rev_orient) {	/* reverse Z-axis to face outside */
		inpfile[i].theta = 180. - inpfile[i].theta;
		inpfile[i].phi = 360. - inpfile[i].phi;
	}
				/* convert to Y-up orientation */
	inpfile[i].phi += 90.-inpfile[i].up_phi;
				/* convert angle to grid position */
	set_grid_pos(&inpfile[i]);
	return(1);
}

/* Load a set of measurements corresponding to a particular incident angle */
static int
add_pabopto_inp(const int i)
{
	FILE		*fp = fopen(inpfile[i].fname, "r");
	double		theta_out, phi_out, val[3];
	int		n, c;
	
	if (fp == NULL || fseek(fp, inpfile[i].dstart, 0) == EOF) {
		fputs(inpfile[i].fname, stderr);
		fputs(": cannot open\n", stderr);
		return(0);
	}
					/* prepare input grid */
	if (!i || cmp_indir(&inpfile[i-1], &inpfile[i])) {
		if (i)			/* process previous incidence */
			make_rbfrep();
#ifdef DEBUG
		fprintf(stderr, "New incident (theta,phi)=(%.1f,%.1f)\n",
					inpfile[i].theta, inpfile[i].phi);
#endif
		if (inpfile[i].nspec)
			set_spectral_samples(inpfile[i].nspec);
		new_bsdf_data(inpfile[i].theta, inpfile[i].phi);
	}
#ifdef DEBUG
	fprintf(stderr, "Loading measurements from '%s'...\n", inpfile[i].fname);
#endif
					/* read scattering data */
	while (fscanf(fp, "%lf %lf %lf", &theta_out, &phi_out, val) == 3) {
		for (n = 1; n < inpfile[i].nspec; n++)
			if (fscanf(fp, "%lf", val+n) != 1) {
				fprintf(stderr, "%s: warning: unexpected EOF\n",
						inpfile[i].fname);
				fclose(fp);
				return(1);
			}
					/* check if scatter angle is too low */
		if (fabs(theta_out - 90.) < lim_graze-FTINY)
			continue;
		if (rev_orient) {	/* reverse Z-axis to face outside? */
			theta_out = 180. - theta_out;
			phi_out = 360. - phi_out;
		}
		phi_out += 90.-inpfile[i].up_phi;
		add_bsdf_data(theta_out, phi_out, val, inpfile[i].isDSF);
	}
	n = 0;
	while ((c = getc(fp)) != EOF)
		n += !isspace(c);
	if (n) 
		fprintf(stderr,
			"%s: warning: %d unexpected characters past EOD\n",
				inpfile[i].fname, n);
	fclose(fp);
	return(1);
}

#ifndef TEST_MAIN

#define	SYM_ILL		'?'		/* illegal symmetry value */
#define	SYM_ISO		'I'		/* isotropic */
#define	SYM_QUAD	'Q'		/* quadrilateral symmetry */
#define	SYM_BILAT	'B'		/* bilateral symmetry */
#define	SYM_ANISO	'A'		/* anisotropic */
#define SYM_UP		'U'		/* "up-down" (180°) symmetry */

static const char	quadrant_rep[16][16] = {
				"in-plane","0-90","90-180","0-180",
				"180-270","0-90+180-270","90-270",
				"0-270","270-360","270-90",
				"90-180+270-360","270-180","180-360",
				"180-90","90-360","0-360"
			};
static const char	quadrant_sym[16] = {
				SYM_ISO, SYM_QUAD, SYM_QUAD, SYM_BILAT,
				SYM_QUAD, SYM_ILL, SYM_BILAT, SYM_ILL,
				SYM_QUAD, SYM_BILAT, SYM_ILL, SYM_ILL,
				SYM_BILAT, SYM_ILL, SYM_ILL, SYM_ANISO
			};

/* Read in PAB-Opto BSDF files and output RBF interpolant */
int
main(int argc, char *argv[])
{
	extern int	nprocs;
	static char	gval_buf[16];
	char *		auto_grazing = NULL;
	const char	*symmetry = "0";
	int		ninpfiles, totinc;
	int		a, i;

	progname = argv[0];			/* get options */
	for (a = 1; a < argc && argv[a][0] == '-'; a++)
		switch (argv[a][1]) {
		case 't':
			rev_orient = !rev_orient;
			break;
		case 'n':
			nprocs = atoi(argv[++a]);
			break;
		case 's':
			symmetry = argv[++a];
			break;
		case 'g':
			if (toupper(argv[a+1][0]) == 'A')
				auto_grazing = argv[a+1] = gval_buf;
			else
				lim_graze = atof(argv[a+1]);
			++a;
			break;
		default:
			goto userr;
		}
	totinc = ninpfiles = argc - a;		/* initialize & sort inputs */
	if (ninpfiles < 2)
		goto userr;
	if (toupper(symmetry[0]) == SYM_UP)	/* special case for "up" symmetry */
		totinc += ninpfiles;
	inpfile = (PGINPUT *)malloc(sizeof(PGINPUT)*totinc);
	if (inpfile == NULL)
		return(1);
	if (auto_grazing)
		lim_graze = 90.;
	for (i = 0; i < ninpfiles; i++) {
		if (!init_pabopto_inp(i, argv[a+i]))
			return(1);
		if (auto_grazing && fabs(inpfile[i].theta - 90.) < lim_graze)
			lim_graze = fabs(inpfile[i].theta - 90.);
	}
	if (auto_grazing)
		sprintf(auto_grazing, "%.2f", lim_graze);
	for (i = ninpfiles; i < totinc; i++) {	/* copy for "up" symmetry */
		inpfile[i] = inpfile[i-ninpfiles];
		inpfile[i].phi += 180.;		/* invert duplicate data */
		inpfile[i].up_phi -= 180.;
		set_grid_pos(&inpfile[i]);	/* grid location for sorting */
	}
	qsort(inpfile, totinc, sizeof(PGINPUT), cmp_indir);
						/* compile measurements */
	for (i = 0; i < totinc; i++)
		if (!add_pabopto_inp(i))
			return(1);
	make_rbfrep();				/* process last data set */
						/* check input symmetry */
	switch (toupper(symmetry[0])) {
	case '0':				/* unspecified symmetry */
		if (quadrant_sym[inp_coverage] != SYM_ILL)
			break;			/* anything legal goes */
		fprintf(stderr, "%s: unsupported phi coverage (%s)\n",
				progname, quadrant_rep[inp_coverage]);
		return(1);
	case SYM_UP:				/* faux "up" symmetry */
		if (quadrant_sym[inp_coverage] == SYM_ANISO)
			break;
		/* fall through */
	case SYM_ISO:				/* usual symmetry types */
	case SYM_QUAD:
	case SYM_BILAT:
	case SYM_ANISO:
		if (quadrant_sym[inp_coverage] == toupper(symmetry[0]))
			break;			/* matches spec */
		fprintf(stderr,
		"%s: phi coverage (%s) does not match requested '%s' symmetry\n",
				progname, quadrant_rep[inp_coverage], symmetry);
		return(1);
	default:
		fprintf(stderr,
  "%s: -s option must be Isotropic, Quadrilateral, Bilateral, Up, or Anisotropic\n",
				progname);
		return(1);
	}
#ifdef DEBUG
	fprintf(stderr, "Input phi coverage (%s) has '%c' symmetry\n",
				quadrant_rep[inp_coverage],
				quadrant_sym[inp_coverage]);
#endif
	build_mesh();				/* create interpolation */
	SET_FILE_BINARY(stdout);		/* start header */
	newheader("RADIANCE", stdout);
	printargs(argc, argv, stdout);
	fputnow(stdout);
	save_bsdf_rep(stdout);			/* complete header + data */
	return(0);
userr:
	fprintf(stderr, "Usage: %s [-t][-n nproc][-s symmetry][-g angle|'A'] meas1.dat meas2.dat .. > bsdf.sir\n",
					progname);
	return(1);
}

#else		/* TEST_MAIN */

/* Test main produces a Radiance model from the given input file */
int
main(int argc, char *argv[])
{
	PGINPUT	pginp;
	char	buf[128];
	FILE	*pfp;
	double	bsdf, min_log;
	FVECT	dir;
	int	i, j, n;

	progname = argv[0];
	if (argc != 2) {
		fprintf(stderr, "Usage: %s input.dat > output.rad\n", progname);
		return(1);
	}
	inpfile = &pginp;
	if (!init_pabopto_inp(0, argv[1]) || !add_pabopto_inp(0))
		return(1);
						/* reduce data set */
	if (make_rbfrep() == NULL) {
		fprintf(stderr, "%s: nothing to plot!\n", progname);
		exit(1);
	}
#ifdef DEBUG
	fprintf(stderr, "Minimum BSDF = %.4f\n", bsdf_min);
#endif
	min_log = log(bsdf_min*.5 + 1e-5);
#if 1						/* produce spheres at meas. */
	puts("void plastic yellow\n0\n0\n5 .6 .4 .01 .04 .08\n");
	n = 0;
	for (i = 0; i < grid_res; i++)
	    for (j = 0; j < grid_res; j++)
		if (dsf_grid[i][j].sum.n > 0) {
			ovec_from_pos(dir, i, j);
			bsdf = dsf_grid[i][j].sum.v /
			   ((double)dsf_grid[i][j].sum.n*output_orient*dir[2]);
			if (bsdf <= bsdf_min*.6)
				continue;
			bsdf = log(bsdf + 1e-5) - min_log;
			ovec_from_pos(dir, i, j);
			printf("yellow sphere s%04d\n0\n0\n", ++n);
			printf("4 %.6g %.6g %.6g %.6g\n\n",
					dir[0]*bsdf, dir[1]*bsdf, dir[2]*bsdf,
					.007*bsdf);
		}
#endif
#if 1						/* spheres at RBF peaks */
	puts("void plastic red\n0\n0\n5 .8 .01 .01 .04 .08\n");
	for (n = 0; n < dsf_list->nrbf; n++) {
		RBFVAL	*rbf = &dsf_list->rbfa[n];
		ovec_from_pos(dir, rbf->gx, rbf->gy);
		bsdf = eval_rbfrep(dsf_list, dir);
		bsdf = log(bsdf + 1e-5) - min_log;
		printf("red sphere p%04d\n0\n0\n", ++n);
		printf("4 %.6g %.6g %.6g %.6g\n\n",
				dir[0]*bsdf, dir[1]*bsdf, dir[2]*bsdf,
				.011*bsdf);
	}
#endif
#if 1						/* output continuous surface */
	puts("void trans tgreen\n0\n0\n7 .7 1 .7 .04 .04 .9 1\n");
	fflush(stdout);
	sprintf(buf, "gensurf tgreen bsdf - - - %d %d", grid_res-1, grid_res-1);
	pfp = popen(buf, "w");
	if (pfp == NULL) {
		fprintf(stderr, "%s: cannot open '| %s'\n", progname, buf);
		return(1);
	}
	for (i = 0; i < grid_res; i++)
	    for (j = 0; j < grid_res; j++) {
		ovec_from_pos(dir, i, j);
		bsdf = eval_rbfrep(dsf_list, dir);
		bsdf = log(bsdf + 1e-5) - min_log;
		fprintf(pfp, "%.8e %.8e %.8e\n",
				dir[0]*bsdf, dir[1]*bsdf, dir[2]*bsdf);
	    }
	if (pclose(pfp) != 0)
		return(1);
#endif
	return(0);
}

#endif		/* TEST_MAIN */
